fvtest/compilerunittest/p/BinaryEncoder.cpp

```diff
@@ -0,0 +1,2856 @@
+
+TEST_P(PPCTrg1Src3EncodingTest, encode) {
+    auto instr = generateTrg1Src3Instruction(
+        cg(),
+        std::get<0>(GetParam()),
+        fakeNode,
+        cg()->machine()->getRealRegister(std::get<1>(GetParam())),
+        cg()->machine()->getRealRegister(std::get<2>(GetParam())),
+        cg()->machine()->getRealRegister(std::get<3>(GetParam())),
+        cg()->machine()->getRealRegister(std::get<4>(GetParam()))
+    );
+
+    ASSERT_EQ(
+        std::get<5>(GetParam()),
+        encodeInstruction(instr)
+    );
+}
+
+TEST_P(PPCTrg1Src3EncodingTest, encodeRecordForm) {
+    TR::InstOpCode opCode = std::get<0>(GetParam());
+
+    if (!opCode.hasRecordForm())
+        return;
+
+    TR::InstOpCode recordOpCode = opCode.getRecordFormOpCodeValue();
+    auto flippedBits = findFlippedBits(opCode, recordOpCode);
+    auto instr = generateTrg1Src3Instruction(
+        cg(),
+        recordOpCode.getOpCodeValue(),
+        fakeNode,
+        cg()->machine()->getRealRegister(std::get<1>(GetParam())),
+        cg()->machine()->getRealRegister(std::get<2>(GetParam())),
+        cg()->machine()->getRealRegister(std::get<3>(GetParam())),
+        cg()->machine()->getRealRegister(std::get<4>(GetParam()))
+    );
+
+    ASSERT_EQ(
+        std::get<5>(GetParam()) ^ flippedBits,
+        encodeInstruction(instr)
+    );
+}
+
+class PPCFenceEncodingTest : public PowerBinaryEncoderTest, public ::testing::WithParamInterface<uint32_t> {};
+
+TEST_P(PPCFenceEncodingTest, testEntryRelative32Bit) {
+    uint32_t relo = 0xdeadc0deu;
+
+    auto node = TR::Node::createRelative32BitFenceNode(reinterpret_cast<void*>(&relo));
+    auto instr = generateAdminInstruction(cg(), TR::InstOpCode::fence, fakeNode, node);
+
+    ASSERT_EQ(BinaryInstruction(), encodeInstruction(instr, GetParam()));
+    ASSERT_EQ(GetParam(), relo);
+}
+
+template <typename T>
+class MakeVector {
+    std::vector<T> _vals;
+
+    void add_vals() {}
+
+    template <typename... Ts>
+    void add_vals(T next_val, Ts... more_vals) {
+        _vals.push_back(next_val);
+        add_vals(more_vals...);
+    }
+public:
+    template <typename... Ts>
+    MakeVector(Ts... vals) {
+        add_vals(vals...);
+    }
+
+    const std::vector<T>& operator*() const {
+        return _vals;
+    }
+};
+
+/**
+ * Note that the expected binary encoding results for these tests are hardcoded here. In most
+ * cases, these values are generated by compiling an assembly file containing the test instructions
+ * using GCC, then using objdump -d to disassemble the resulting object file.
+ *
+ * Unfortunately, due to the fact that the way we represent instructions is different from the way
+ * that they're represented in the assembly expected by GCC (especially instructions like rldicl
+ * and rlwinm), it is not possible to fully automate this process.
+ */
```

Comment:
Could you elaborate how you found object files with a particular instruction? For example I want to test encoding of instruction _foo_, how do I find an object file which contains _foo_? Did you manually make use of _foo_ or dumped disassembly of some large binary and hoped to find _foo_ in there, or?

Would like to know the process because I'd like to make sue of this infrastructure to test the other code generators as well.

Reply:
What I did was use a very basic Python script (only ~30 lines) to output a custom assembly file using each of a number of instructions with the required registers/immediates (I manually designated these for each group of instructions to test), run `gcc` on that assembly file to produce an object file, parse the output of `objdump -d` to read the binary encoded instructions from that object file, then print out the lines that need to be pasted into the test file.

This script has very little in the way of actual logic and it basically just contains a couple loops with some string interpolation to print things and a regex to parse the objdump output (this would need significant modification to be able to process variable-length CISC instructions). I've had to make modifications for many different instructions due to some oddities in how the Power ISA wants instructions to be formatted and for the different combinations of registers/immediates that need to be tested for different kinds of instructions, so it's not very general at the moment.